### ISSUE TITLE
zathura compatibility with latexmk setting

### ISSUE THREAD

**Issue by chakravala** [lervag/vimtex#1063]
On arch linux using latest vim/gvim.

When I launch vimtex, this is my command for it in `.bashrc`
```bash
alias vimtex='vim --servername vim'
```
the configuration is `.vimrc`
```vim
let g:vimtex_view_method = 'zathura'
```

This configuration works fine for me, except that when I launch it and press `\ll` for the first time, zathura is opened and the document is displayed in its previous configuration and then the program freezes for a few seconds, and then it resets the display on zathura with the horizontal scroll reset and a bunch of view settings reset.

However, if I add this to `.vimrc`
```vim
let g:vimtex_compiler_progname = 'latexmk'
```
then the issue goes away and when I press `\ll` the zathura imediately displays the document and does not freeze for a few seconds and **does not** reset my view settings or horizontal scroll config.

**However** now with this setting I can no longer ctrl+click on zathura to communicate back to vimtex where on what line I want the editing cursor to be placed.

How can I make this work using this option? There's probably some simple thing I am overlooking.

Thanks for any assistance.

**Comment by lervag** [comment]
First,

```vim
let g:vimtex_compiler_progname = 'latexmk'
```

does not make any sense. This option is meant to define the name/path to the proper Vim executable for doing remote server/client calls during callbacks.

I suspect that the `.latexmkrc` is relevant here. Could you check if the problem persists if you remove that line from your `.latexmkrc` file?

**Comment by chakravala** [comment]
Then why is it that when I use
```vim
let g:vimtex_compiler_progname = 'latexmk'
```
this solves the issue with zathura resetting my view settings, but I don't really understand what's going on with it. then what is the setting that is supposed to be used for it? It doesn't seem to matter what I enter into the value for that.

Also, the `.latexmkrc` setting does not affect the issue at all, removing it or keeping it doesn't change it. Does it have something to do with how to call pdflatex using synctex?

**Comment by lervag** [comment]
Ok, since `.latexmkrc` is not relevant I'm removing all mentions of it from the thread.

I'm thinking the reason it solves the issue is because the default of `g:vimtex_compiler_progname` is the full path to gvim, which starts with `C:\Program Files`. The problem here is related to escaping during the callbacks from latexmk.

Could you please update the original post to better match the issue template?

**Comment by chakravala** [comment]
Okay, seems you already removed it from the post. I'm on arch linux, using the latest vim/gvim, I'm not sure what additional information to provide, since the other settings I use don't affect this issue.

Could you explain more about the escaping during callbacks from latexmk?

**Comment by lervag** [comment]
Ok, so: Please provide a minimal working example, that is, a minimal vimrc file, a minimal latex file, and the steps to reproduce.

An example:

1. `minivimrc`:

    ```vim
    set nocompatible
    let &rtp  = '~/.vim/bundle/vimtex,' . &rtp
    let &rtp .= ',~/.vim/bundle/vimtex/after'
    filetype plugin indent on
    syntax enable

    let g:vimtex_view_method = 'zathura'
    ```

2. A minimal `.tex` file:

    ```tex
    \documentclass{minimal}
    \begin{document}

    Hello World!

    \end{document}
    ```

3. Now do `vim --servername VIM -u minivimrc mwe.tex`.

4. Do `\ll`

    - expect: _some behaviour_
    - observe: _some other behaviour_

**Comment by lervag** [comment]
btw, sorry for mistaking that you were on Windows. Your issue is probably something else. But it is _crucial_ that you describe it in enough detail that I might reproduce it.

**Comment by chakravala** [comment]
Yes, I created the example files, and it reproduces my issue, and if you change the `g:vimtex_compiler_progname` it loads the zathura without reseting as described above.

Since this is related to the flickering issue I had before, I'm guessing you are not able to reproduce it

Is my zathura configuration at all relevant?
```
set synctex true
set synctex-editor-command "vim --servername VIM --remote-silent +%2<Enter>zz %1"
```

**Comment by chakravala** [comment]
Okay, well it seems if I put
```
let g:vimtex_compiler_progname = '/usr/bin/vim`
```
Then it gives me the same result as completely omitting that setting

**Comment by chakravala** [comment]
What happens is that the VIM screen goes completely blank for a few seconds, using the default setting, and then when it finishes processing (not sure what is processing), it redraws the screen and then resets the document view for zathura. This doesn't occur if I change the `compiler_progname` to something invalid, but zathura is still able to update when I make changes in vim, the only issue it causes is that zathura is no longer able to communicate back.

**Comment by lervag** [comment]
Ok, thanks. I'm pushing some test files now, and the commit should appear in this thread. Please review the test files and confirm that they represent your minimal example.

Also, please give me the output of `\li` after you have opened a tex file.

**Comment by lervag** [comment]
So, I'm curious, what happens if you remove the `set synctex-editor` line (or comment it out) from the `zathurarc` file.

Also, it would help if I understood what exactly is happening. As you probably understand, I can't reproduce your issue (at least I don't think so). Perhaps it is possible to take a screenshot or screencast?

**Comment by chakravala** [comment]
I have emailed you screenshots now, and the `set synctex-editor` does not affect the issue outcome

**Comment by lervag** [comment]
Ok, I've looked through the screenshots.

Could you check if the problem persists if you set the following option?

```vim
let g:vimtex_compiler_latexmk = { 
      \  'callback' : 0,
      \}
```

Note, please try this with the default `g:vimtex_compiler_progname`.

**Comment by lervag** [comment]
Also, some more thoughts and questions:

- Which version of Zathura are you using?
- Can you reproduce the problem with only the `\lv` mapping for opening the viewer?
- Can you reproduce the problem with the test files that I pushed? That is, if you do

    1. `cd ~/.vim/bundle/vimtex/test/issues/1063` (or wherever you have installed vimtex)
    2. `vim -u minivimrc` or `vim --servername vim -u minivimrc`
    3. `\ll`
    4. Repeat steps 2 and 3 after editing `minivimrc` and changing the `g:vimtex_compiler_progname` option.

**Comment by chakravala** [comment]
Yep, that solves it, with `g:vimtex_compiler_latexmk` you have.

**Comment by lervag** [comment]
What do you mean? Which setting solves it? Is it this?

```vim
if has('nvim')
  let g:vimtex_compiler_progname = 'nvr'
endif
```

If so, then it is because you are running neovim, not vim or gvim. And this is **essential** information that you never told me.

In any case, I am happy that things are working!

**Comment by chakravala** [comment]
This one solves it, that is the only change it needed, now everything works without any issue.
```vim
let g:vimtex_compiler_latexmk = { 
      \  'callback' : 0,
      \}
```
With this setting, the resetting of the view no longer occurs, and there is no longer a wait time before vim becomes usable, and vim does not show any blank screen for a few seconds anymore.

So I'm not running neovim. Also, my zathura version is the latest.

Also, with `vimtex_compiler_progname` it doesn't work, it had to be `vimtex_compiler_latexmk`.

**However,**

Hmm but when I use `\lv` the zathura view settings are not restored properly. The communication from zathura to vimtex works, but when it initially opens zathura, the PDF view is reset for `\lv`. But I noticed this issue with the resetting of the view on `\lv` only occurs if I do  not move the text editing position cursor in vim past the line for `\begin{document}`. That makes sense, since the view cannot be set if it is before the document starts.

NOW I understand. If I go back to the default setting of `vimtex_compiler_latexmk`, then If I use the `\lv` or `\ll` command while the editing cursor is before `\begin{document}`, then it does the resetting of the PDF viewer thing. However, with the new setting provided above, this does not occur anymore, and if I press `\lv` or `\ll` after I move the cursor below `\begin{document}`, then the screen still goes blank for a few seconds and I have to wait, but the view is not reset, instead it is set to wherever the cursor was.

So the new setting solves the issue though, there is no more blank screen for a few seconds, and if I press `\ll` at the very first line before `\begin{document}`, then zathura is immediately opened with my previous view settings (not reset) and it is not scrolled to some specific place, but wherever I had it when I last had it open (which is the behavior I wanted).

Thanks a lot again. Hope this information clarifies more about what's going on.

**Comment by lervag** [comment]
`g:vimtex_compiler_progname` and `g:vimtex_compiler_latexmk` are two _different_ options that do _differnet_ things.

What specifically is _latest_? 0.3.8?

So: Setting the `g:vimtex_compiler_latexmk` option is not a solution. It disables a very convenient callback feature from latexmk that e.g. opens the quickfix buffer on errors, shows compilation status, etc. I asked you to do it because it will also prevent zathura from being opened by the callback directly.

For me, it seems you are still not describing the issue in such a way that it is clear to me what the problem is. But if you are happy with the current setting, then I am happy. But I am using the same system as you with Zathura, and for me everything works as expected with no delays and no strange blank screens.

Do you have `xdotool` installed?

**Comment by chakravala** [comment]
Yes, 0.3.8 and I have `xdotool` installed also. Hmm, well not having the errors and compilation status would be inconvenient, so I would like to have that. Guess it's not entirely solved yet, but at least it's more clear now where the issue is coming from.

With the default setting, the main issue I have is that when I do `\ll` the vim screen will go blank for a few seconds, and then when the vim screen is restored, vim seems to essentially run the `\lv` ommand and proceeds to set the view to highlight where my cursor is. So with the default setting, there is some kind of process running that makes the vim screen go blank for a few seconds, and then calls the `\lv` command which causes the view to reset. That's what I believe is happening.

**Comment by lervag** [comment]
Ok, I have two ideas. First, could you do the following steps:

    0. Remove your latexmk and Zathura settings (don't delete, only move away so they are not active.)
    1. `cd ~/.vim/bundle/vimtex/test/issues/1063` (or wherever you have installed vimtex)
    2. `vim -u minivimrc` or `vim --servername vim -u minivimrc` or `gvim -u minivimrc` (note that the `minivimrc` loads the `minimal.tex` file implicitly).
    3. `\ll`
    4. Report whether you still have the issue. The point is to check if the issue can be caused by configuration.

Second: Are you using Wayland? Which window manager/desktop manager are you using?

**Comment by chakravala** [comment]
Alright, I did do steps 0,1,2,3, so here is 4: in order to be able to say whether the issue is reproduced, I have modified the `minimal.tex` file as follows:
```TeX
\documentclass{minimal}
\begin{document}

Hello World! \newpage
Hello again \newpage
once more!

\end{document}
```
Then doing the test, I have found that using `minivimrc` that you provided the issue still happened i.e., the vim screen went blank on pressing `\ll` and after the screen is restored from being blank the zathura window is reset as if the `\lv` command has been executed (but I pressed `\ll`).

No, not on Wayland yet, I'm on X with i3 window manager. However, I intend to switch to Wayland at some point.

**Comment by chakravala** [comment]
Also, I'm not sure if this is how it should be or not, but from the other issue, you suggested to check
```shell
[flow@research ~]$ ps aux|grep latexmk
flow      9101  0.0  0.0  10884  2224 pts/4    S+   09:08   0:00 grep latexmk
```
So is that a spurious process, or is it correct to have one of those running, when I don't have any vim and zathuras open?

**Comment by lervag** [comment]
No, it does not seem like there is a spurious process, the process in that list is the `grep` process. But obviously, there should be only one latexmk process per project, and if you want to control things from Vim, then the process should be the one started from Vim. If you have two latexmks running on the same project, things are going to get nasty.

In your previous post you write "you have to relaunch vim and then press `\ll` as follows:", but what follows does not seem connected...

**Comment by chakravala** [comment]
When I'm running vim, I only have 1 instance per vim, I just checked.

In the previous post, all I meant was that before you press `\ll` to do the test, you have to compile the document, view it in zathura, and set the view to page 2 or page 3.

Then when you quit zathura and vim, relaunch vim and press `\ll`. Then zathura launches, it goes to the same view it had when you last quit (page 2 or page 3), but the vim screen goes blank for a few seconds, and when vim displays the TeX code once again, the zathura view is reset to the first line of the document, as if the `\lv` command has been initiated.

**Comment by lervag** [comment]
Ok. So, I think we have two issues:

1. The vim screen goes blank for 2-3 seconds.
2. Zathura view resets.

Personally, I use `g:vimtex_view_automatic = 1` because I don't want the viewer to open unless I ask it to. I.e., `\ll` won't open the viewer with that setting. Obviously, this is not a solution, but perhaps this "workflow" also works for you?

So, the reason it takes some time (i.e. 1) is because there is a loop in the Zathura viewer code (in vimtex) to search for an instance of zathura that was started by latexmk. If no such instance is found (after 30*50 ms = 1500 ms), then vimtex starts the viewer itself as if you do `\lv`. I think the issue has to do with this. That is, I am guessing that

1. latexmk starts zathura
2. vimtex through use of xdotool can not find the xwin id and thus after 1,5s it opens a new viewer

Can you try this:

1. Open the minimal example as explained before
2. Do `\lv`
3. Do `\li` and copy the content here, in particular, I am interested in `viewer` part which contains the xwin id. It should be nonzero.

**Comment by chakravala** [comment]
Sorry for the late response, was having internet trouble. Here is `\li`
```
System info
  OS: Linux 4.15.5-1-ARCH
  Vim version: VIM 8.0 (1-1542)
  Has clientserver: true
  Servername: VIM1

vimtex project: minimal
  base: minimal.tex
  root: /home/flow/.vim/bundle/vimtex/test/issues/1063
  tex: /home/flow/.vim/bundle/vimtex/test/issues/1063/minimal.tex
  out: /home/flow/.vim/bundle/vimtex/test/issues/1063/minimal.pdf
  log: /home/flow/.vim/bundle/vimtex/test/issues/1063/minimal.log
  aux: /home/flow/.vim/bundle/vimtex/test/issues/1063/minimal.aux
  fls: /home/flow/.vim/bundle/vimtex/test/issues/1063/minimal.fls
  document class: minimal
  compiler: latexmk
    backend: jobs
    output: /tmp/vS7Yike/1
    configuration: 
      continuous: 1
      callback: 1
      latexmk options:
        -verbose
        -pdf
        -file-line-error
        -synctex=1
        -interaction=nonstopmode
    job: process 24687 run
    cmd: max_print_line=2000 latexmk -verbose -pdf -file-line-error -synctex=1 -interaction=nonstopmode -pvc -e '$new_viewer_always = "0"' -e '$pdf_previewer = "zathura  -x \"/usr/bin/vim --servername VIM1 --remote +\%{line} \%{input}\" \%S"' -e '$success_cmd = "'""'/usr/bin/vim'""' --servername 'VIM1' --remote-expr \"vimtex\#compiler\#callback(1)\""' -e '$failure_cmd = "'""'/usr/bin/vim'""' --servername 'VIM1' --remote-expr \"vimtex\#compiler\#callback(0)\""' 'minimal.tex'
  viewer: Zathura
    xwin id: 48234499
    process: 
      pid: -
      cmd: zathura -x "/usr/bin/vim --servername VIM1 --remote-expr \"vimtex#view#reverse_goto(%{line}, '%{input}')\"" --synctex-forward 1:1:'/home/flow/.vim/bundle/vimtex/test/issues/1063/minimal.tex'  '/home/flow/.vim/bundle/vimtex/test/issues/1063/minimal.pdf' >/dev/null 2>&1 &
  qf: LaTeX logfile
    config: 
      fix_paths: 1
      packages: 
        default: 1
      default: 1
```
So I'm okay with the idea of using `g:vimtex_view_automatic = 1`, but I would prefer if there exists some additional other command aside from `\lv` for opening zathura. I would like to be able to open zathura without highlighting the current line, is that possible? Maybe using `\lo` or something?

For the other issue, is it possible to turn the 30*50 setting into a modifiable option? I'd like to be able to experiment with a setting that works for me. So does that mean there are 30 tries of 50 ms? It might still work with less tries too.

To summarize, having a command to open zathura without the `\lv` behavior of highlighting the current line would be nice and the option to change the number of tries for xdotool. This would make things a very nice workflow for me, if it is not too much trouble to add these features?

Thanks for all the help you've given so far!

**Comment by lervag** [comment]
Sorry, I meant to use `g:vimtex_view_automatic = 0`, not 1. That is, to prevent Zathura from opening automatically.

Note that `\lv` only highlights the current line the first time it is called, but you can disable this with the `g:vimtex_view_forward_search_on_start` setting (i.e. set it to 0 to disable). When the pdf is updated, the line is not highlighted any more. You can also redraw the Zathura window with the `R` map inside Zathura to remove the highlighting. If Zathura is opened and you do `\lv` you perform a forward search.

I hope you can accept that I don't add a new viewer command, nor a new option for the delay (30*50). I don't understand why this does not work as expected on your end, but my suspicion is that there is something strange going on with the xdotool window id search. In any case, I think that the settings

```vim
let g:vimtex_view_automatic = 0
let g:vimtex_view_forward_search_on_start = 0
```

might get you close to what you want.

**Comment by chakravala** [comment]
Thanks again, using those settings I can get the type of behavior I wanted, it just wasn't fully clear to me how I can get the behavior I wanted, but it's all solved now, I don't think there is anything else troubling me. With the two settings you provided, I can do `\lv` to open the zathura, and it now it doesn't highlight the line on the first launch and there also isn't a long delay time using `\lv`, so that solves both my issues.

**Comment by lervag** [comment]
I'm happy it works :)

**Comment by chakravala** [comment]
Hi, so I experimented with the `zathura.vim` file and figured out more about the causes.

With the following modifications, I only need
```vim
let g:vimtex_view_forward_search_on_start = 0
```
and am using the default for `g:vimtex_view_automatic`.

In the following function I commented out 2 lines at the bottom
```vim
function! s:zathura.start(outfile) dict " {{{1
  let l:cmd  = 'zathura'
  let l:cmd .= ' -x "' . g:vimtex_compiler_progname
        \ . ' --servername ' . v:servername
        \ . ' --remote-expr '
        \ .     '\"vimtex#view#reverse_goto(%{line}, ''%{input}'')\""'
  if g:vimtex_view_forward_search_on_start
    let l:cmd .= ' --synctex-forward '
          \ .  line('.')
          \ .  ':' . col('.')
          \ .  ':' . vimtex#util#shellescape(expand('%:p'))
  endif
  let l:cmd .= ' ' . g:vimtex_view_zathura_options
  let l:cmd .= ' ' . vimtex#util#shellescape(a:outfile)
  "let self.process = vimtex#process#start(l:cmd)

  "call self.xwin_get_id()
  let self.outfile = a:outfile
endfunction
```
The first one needed to be commented out because with this configuration it causes the zathura window to be opened twice, so to prevent 2 zathura windows from existing, I had to comment that one, and then the second line is commented in order to prevent vim screen from turning blank and waiting for a few seconds before it becomes usable.

Then I modified the timer in
```vim
function! s:zathura.compiler_callback(status) dict " {{{1
  if !a:status && g:vimtex_view_use_temp_files < 2
    return
  endif

  if g:vimtex_view_use_temp_files
    call self.copy_files()
  endif

  if !filereadable(self.out()) | return | endif

  if g:vimtex_view_automatic
    "
    " Search for existing window created by latexmk
    "   It may be necessary to wait some time before it is opened and
    "   recognized. Sometimes it is very quick, other times it may take
    "   a second. This way, we don't block longer than necessary.
    "
    "if !has_key(self, 'started_through_callback')
    "  for l:dummy in range(30)
    "    sleep 50m
    "    if self.xwin_exists() | break | endif
    "  endfor
    "endif

    if !self.xwin_exists() && !has_key(self, 'started_through_callback')
      call self.start(self.out())
      let self.started_through_callback = 1
    endif
  endif
```
So I have commented the 30*50 milliseconds, to prevent the vim screen from blanking.

With this modification, I can press `\ll` and the document is opened without any wait time or screen blanking and without the highlighting issue, and the inverse search works well.

Would these modifications have any other side effects? With the solution we had yesterday, everything also worked well, but then I had to press `\lv` and then `\ll`, whereas with this modification I only need to press `\ll` once. Also, with the yesterday's configuration, the inverse search did not work properly, since when I click on the PDF, the vim screen was not updated with the new position (but if I move the cursor up or down, then vim jumps to the new position, but just from the click the vim screen was not updated). However, with this modification, when I click on zathura, the vim screen cursor is immediately updated to where it should be on the vim screen.

So those are just some extra things I noticed.

**Comment by lervag** [comment]
Well, these modifications are not going to work in general. What you are doing is essentially disabling the built in Zathura controls. That is, Zathura is now opened by latexmk and vimtex is not trying to detect the xwin id and thus not trying to get control of the situation.

Considering that the Zathura viewer works for me (and as far as I know all other users), I don't think these changes are good.

However, as you are getting into the code, I think it would be more interesting if you could rather go more into depth about _why_ the function `xwin_get_id` makes things go black. This shouldn't happen. This function is defined in `common.vim` in the same folder. You could try to add some `echo`s, `sleep`s and similar to investigate the issue yourself.

To be more specific, I believe the issue is something like this:

- You do `\ll`
- latexmk spawns a Zathura viewer (based on settings passed from vimtex)
- when latexmk has compiled successfully, it calls back to vim with the function `vimtex#compiler#callback`
- this in turn calls `s:zathura.compiler_callback`
- your first delay comes because `self.xwin_exists` always returns false
- I think this is the point where things are not working as they should - since latexmk starts zathura, `self.xwin_exists` should return true at some point when the viewer has opened correctly

Thus, to further debug, I would suggest you look in more depth at `common.vim` line 138. You can do `unsilent echo "test"` at various points in the code (I am not sure if `unsilent` is necessary). This way I think you should be able to find some reason why it always returns false, instead of eventually finding the xwin id of the zathura window and return true.

**Comment by chakravala** [comment]
In the pull request I did not disable the `xwin_get_id` part, that part turned out to be essential and is not the part that costs very much time anyway. That's from before, when I was experimenting.

What causes the time delay is `let self.process = vimtex#process#start(l:cmd)`. So my modification makes it so that zathura is only launched when `\ll` is pressed, it is not launched when `\lv` is pressed, but the `\lv` functionality still works as usual, after `\ll` is used to open zathura.

The advantage of disabling the process start there is improved time performance when zathura is initialized for the first time with `\ll`.

The other modification makes it so that there is no timer that checks for other processes. This is no longer needed, since `\lv` does not initiate a new process anymore.

Together, these two modifications / settings work very well for the behavior I wanted.

And I repeat, I did not disable the `xwin_get_id`, that was only in the comment above. Everything works very well this way, no issues. If you only use one of the modifications, then you will have some issues, but if you use them both together, they work perfectly.

Thanks for the suggestions about how to dig in deeper into the `common.vim`, I will take a look.

**Comment by chakravala** [comment]
@lervag , now I have investigated this some more and found out a stable way to fox it by adding an extra step to the `build_cmd()` function for checking whether the `xwin_id` is valid or not.

You can see the changes [here](https://github.com/lervag/vimtex/compare/master...chakravala:zathura-mod)

Could you please test these modifications with your setup, and confirm that everything is functioning properly, as far as zathura and vimtex and latexmk is concerned?

With this modified `build_cmd()`, the two configuration options are no longer needed, they can actually be made permanent, i.e., the 30*50 timer can be completely removed from the code without affecting the startup behavior or causing issues with launching multiple windows.

I don't know how this will interact with other viewers, but for zathura this should work very well.

Please let me know if you take a look and test it with your setup.

**Comment by lervag** [comment]
I don't have the time to read and respond to everything now, but I have one comment: You can simplify your suggested change quite a bit by instead of removing this:

    \ || get(get(b:vimtex, 'viewer', {}), 'xwin_id') > 0

you change it to something like this:

```vim
    try
      l:xwin_exists = b:vimtex.viewer.xwin_exists()
    catch /.*/
      l:xwin_exists = 0
    endtry
    " ...
    \ || l:xwin_exists
```

I.e., instead of copying the code from the `xwin_exists()` function you simply call that function.

But again, as I said, the current version works as expected for me, and so I don't really see any motivation for changing it. It would be much better if you were able to figure out why things are not working as expected on your end in such a way that I can reproduce your problem. It should be reproducible, since we are both on Arch. Perhaps it is related to our difference in window manager?

**Comment by lervag** [comment]
Ok, I've read your posts, and again: You are saying that the delay comes from the `vimtex#process#start`, but as I've said: I can't reproduce any delays or anything similar to what you are describing. I also know there are other vimtex users who use zathura, and they have never mentioned any issues like this. Thus before I make any changes I want to understand your problem more directly.

**Comment by chakravala** [comment]
This is strange, but for some reason if I call `b:vimtex.viewer.xwin_exists()` there is a longer delay time than if I just simply insert the code into that location directly. So it seems to make a difference for how fast it gets done for some reason. Mind you, I am doing my `vim` work on computer that is not very powerful, I have another computer that is more powerful that I can test it on. But I'd like `vim` to work as fast as possible on my less pwerful computer, which is why I'm trying to figure out why this part of the plugin is slowing down my initialization time.

I'm going to keep experimenting around with it and will post here when I have more findings.

**Comment by lervag** [comment]
Thanks! I appreciate your efforts into investigating this yourself. I know it might not be easy, but I hope you understand that it is very difficult for me when I am not able to replicate your problems.

**Comment by chakravala** [comment]
Alright, so I switched to Wayland, but I get `Viewer cannot find Zathura window ID`, I'm still new to this whole Wayland thing, how do I resolve it?

**Comment by lervag** [comment]
Ah, you should not ask me anything related to Wayland. I won't know, as I've never tried it. And as such, it is very difficult for me to help you here.

But, let's try: Do you get things to work from a terminal, that is, can you get the window ID from a terminal? If you open a PDF with Zathura, then do the following:

```sh
xdotool search --class Zathura
```

What happens?

**Comment by chakravala** [comment]
The nature of the Wayland protocol is that this type of information is no longer accessible to programs, for the sake of increased security. So some kind of other technique is likely needed.

So no, it does not work.

Not sure what window manager you are using, but if you are using `i3` it is fantastically easy to try out wayland without any hassle. The `sway` wayland compositor can read your `i3` configuration, so all you have to do is install `sway` with the package manager and copy your `i3` config to `.config/sway`. Then, if you exit your `X` session, instead of using `startx` launch `sway` from the TTY.

Voila, now you have the choice between `i3` with `X` or `sway` with `Wayland` with no hassle. Don't know how easy it is for other window managers.

But I have not figured out yet how to get the window class ID information in Wayland in general yet.

**Comment by lervag** [comment]
I'm using [fluxbox](http://fluxbox.org/) and I've not looked into Wayland myself yet. As such, I can't really help with that at all.

But for Xorg, things should work. And with X, I think you should be able get a window class ID. Can you confirm that?

**Comment by chakravala** [comment]
Yes, on X it works fine. So I am sticking with X for my main work for now, but it's good to also have the wayland available, the only issue I see with the wayland is obtaining the window class ID.

**Comment by lervag** [comment]
It seems, as you stated, that xdotool and similar tools will not be available on Wayland for quite some time. Relatively recent posts on [superuser](https://superuser.com/questions/1032270/wayland-alternative-for-xorgs-xdotool) and [askubuntu](https://askubuntu.com/questions/956640/equivalent-to-xdotool-for-wayland) states that Wayland does not support xdotool like behaviour due to security reasons, and that Wayland compositors might implement such features in the feature.
